### PR TITLE
Add Let's Encrypt SSL Support

### PR DESCRIPTION
**Pull request by Daviey**
Let's Encrypt provides free TLS/SSL certificates, which is trusted by
most platforms [ letsencrypt.org ].

Signed-off-by: Dave Walker (Daviey) <email@daviey.com>

### CODE REVIEW

**Comment by Daviey** [comment]
The commit "Add Mozilla's CA Cert support to curl", has a separate pull request tracked at https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/pull/483

If https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/pull/483 lands, then there are only two commits on this.

Whilst this change seems to WFM, i think it would benefit from other people testing.

Thanks

**Comment by jmtatsch** [comment]
It won't break the current self-signed certificates, right?

**Comment by redi-walkerd** [comment]
@jmtatsch No, won't break existing setups.  On first boot (run.sh) creates the ssl cert, as it doesn't already exist.

*IF* a user wants to use LetsEncrypt, they need to run config/ssl/letsencrypt_setup.sh by hand, and that will replace the self-signed cert.

On next boot, the cert won't be touched by run.sh - as it then already exists.

File: firmware_mod/config/ssl/letsencrypt_setup.sh [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+./acme.sh --issue -d ${SSL_DOMAIN} --home ${ACMEPATH} \
+          -w ${CONFIGPATH}/../www/ \
+          ${DEBUG}
+
+./acme.sh --install-cert -d ${SSL_DOMAIN} --home ${ACMEPATH} \
+          --cert-file ${ACMEPATH}/host.crt \
+          --key-file  ${ACMEPATH}/host.key \
+          --fullchain-file ${ACMEPATH}/fullchain.crt \
+          --reloadcmd  "cat ${ACMEPATH}/host.crt ${ACMEPATH}/host.key > ${CONFIGPATH}/lighttpd.pem ;\
+                        pkill lighttpd.bin ;\
```
[the remark this answers is not in the document]
Comment (redi-walkerd) [reply firmware_mod/config/ssl/letsencrypt_setup.sh]: Good thinking, i'll try this.

File: firmware_mod/config/ssl/letsencrypt_setup.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+DEBUG="--test --debug"
+
+CONFIGPATH="/system/sdcard/config"
+ACMEPATH="${CONFIGPATH}/ssl/acme"
+SSL_DOMAIN_PATH="${CONFIGPATH}/ssl_domain.conf"
+
+if [ ! -f "acme.sh" ]; then
+  echo "- Downloading the acme.sh script..."
+  curl -s https://raw.githubusercontent.com/Neilpang/acme.sh/master/acme.sh > acme.sh
+fi
```
[the remark this answers is not in the document]
Comment (redi-walkerd) [reply firmware_mod/config/ssl/letsencrypt_setup.sh]: yeah, fair point.

**Comment by redi-walkerd** [comment]
@tssva Yeah.. I get your point about this only working with a single camera, and needing to expose :80.. but DNS has to be either manual or using a nameserver that has a supported API.

This script i feel is a reasonable contrib script that can be opted into..  I don't really want to get into the DNS support myself TBH.

Perhaps this should land as a pointer to other people, which can take it further?  Getting this far already required fixing busybox and making changes to the default cron!

**Comment by Daviey** [comment]
@tssva Yeah.. I get your point about this only working with a single camera, and needing to expose :80.. but DNS has to be either manual or using a nameserver that has a supported API.

This script i feel is a reasonable contrib script that can be opted into.. I don't really want to get into the DNS support myself TBH.

Perhaps this should land as a pointer to other people, which can take it further? Getting this far already required fixing busybox and making changes to the default cron!


@jmtatsch No, won't break existing setups. On first boot (run.sh) creates the ssl cert, as it doesn't already exist.

IF a user wants to use LetsEncrypt, they need to run config/ssl/letsencrypt_setup.sh by hand, and that will replace the self-signed cert.

On next boot, the cert won't be touched by run.sh - as it then already exists.

File: firmware_mod/config/ssl/letsencrypt_setup.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (Daviey) [reply firmware_mod/config/ssl/letsencrypt_setup.sh]: Good thinking!  I tried this, but it didn't work as might have expected:

**Comment by jmtatsch** [comment]
Seems to be a solid starting point for future letsencrypt endeavours.
Thank you for your contribution.

**Pull request merged**